repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Authorization.scala

```diff
@@ -38,30 +38,44 @@ object KeystoneV2Authorization extends LazyLogging {
-      authorizeEndpoints(config.getRequireServiceEndpoint, userIsPreAuthed, endpoints)
-    }
-    authResult match {
-      case Success(_) => AuthorizationPassed(scopedRolesToken, matchedTenant)
-      case Failure(exception) => AuthorizationFailed(scopedRolesToken, matchedTenant, exception)
+  // TODO: Break this into separate authorizeTenants/authorizeEndpoints calls to separate concerns
+  def doAuthorization(config: KeystoneV2Config, request: HttpServletRequestWrapper, tenantToRolesMap: TenantToRolesMap, endpoints: => Try[EndpointsData]): AuthorizationInfo = {
+    // NOTE TO REVIEWER: Behavior change -- pre-authorization is determined based on all roles, not just those matching the tenant (since that is part of authorization).
+    //                   This change actually makes the behavior consistent with documentation.
+    if (isUserPreAuthed(config.getPreAuthorizedRoles, tenantToRolesMap.values.flatten.toSeq)) {
+      AuthorizationPassed(tenantToRolesMap, Set.empty)
+    } else {
+      val validateTenant = Option(config.getTenantHandling.getValidateTenant)
+      val tenantsToMatch = validateTenant
+        .map(getRequestTenants(_, request))
+        .getOrElse(Set.empty)
+      val scopedTenantToRolesMap = validateTenant
+        .map(getScopedTenantToRolesMap(_, tenantsToMatch, tenantToRolesMap))
+      // val scopedRoles = scopedTenantToRolesMap.getOrElse(tenantToRolesMap).values.flatten.toSet
+      val matchedTenants = (scopedTenantToRolesMap.getOrElse(Map.empty) - DomainRoleTenantKey).keySet
+      val tenantAuthorization = validateTenant
+        .map(authorizeTenant(_, tenantsToMatch, matchedTenants))
+        .getOrElse(Success(Unit))
+
+      val endpointAuthorization = Option(config.getRequireServiceEndpoint)
+        .map(requiredEndpoint => authorizeEndpoints(requiredEndpoint, endpoints))
+        .getOrElse(Success(Unit))
+
+      tenantAuthorization.flatMap(_ => endpointAuthorization) match {
+        case Success(_) => AuthorizationPassed(scopedTenantToRolesMap.getOrElse(tenantToRolesMap), matchedTenants)
+        case Failure(e) => AuthorizationFailed(scopedTenantToRolesMap.getOrElse(tenantToRolesMap), matchedTenants, e)
+      }
     }
   }
 
-  // TODO: Switch to getRequestTenants (plural) and include all tenants extracted. This is a behavior change.
-  def getRequestTenant(config: ValidateTenantType, request: HttpServletRequestWrapper): String = {
-    Option(config).flatMap(validateTenantConfig =>
-      (validateTenantConfig.getUriExtractionRegexAndHeaderExtractionName.asScala.toStream map {
+  // TODO: Rather than throwing an exception, this method should return a Try.
+  // NOTE: Throwing an exception currently works because this method is always called in a [[Try]] stack.
+  @throws[UnparsableTenantException]
+  def getRequestTenants(config: ValidateTenantType, request: HttpServletRequestWrapper): Set[String] = {
+    Option(config).map(validateTenantConfig =>
```

Comment:
This option is unnecessary, because you are calling this method on a map of an option of the same config.

Reply:
Dangit. Yeah, I changed the way we were handling the config so that every method didn't have to wrap it in an `Option`, but apparently I missed this. Will fix.